### PR TITLE
[#109] ultrasound imaging examples

### PR DESCRIPTION
**Pull request by charlesincharge**
Github Issue: https://github.com/agencyenterprise/neurotechdevkit/issues/109

#### Introduction

In addition to neurostimulation, ultrasound is frequently used for imaging as well. The most common type of ultrasound imaging is diagnostic/anatomical ultrasound imaging, sometimes called B-mode imaging. We add B-mode imaging functionality and examples here: to provide teaching materials, showcase the flexibility of NDK, and inspire potential future directions.

#### Changes

Ultrasound imaging consists of 3 steps:
1. transmit pulse
2. receive echo
3. reconstruct image

Parts 1 and 2 are already implemented in NDK; we add `demodulate.py` and `beamform.py` to implement the image reconstruction steps.

We also update the `ndk.scenario` module to enable receiver locations (not just at the wavefield grid points).

#### Behavior

New functionality.
Tests should probably be added for the new libraries. We can compare against MUST as a reference implementation, similar to: https://app.reviewnb.com/agencyenterprise/bci-experiments/blob/master/experiments/neurotechdevkit-109-ultrasound-imaging/109-beamforming.ipynb/. Open to any other ideas!

Note: currently checking this in as a Jupyter notebook, because that seems easier to attach output to and view. Will repoint this to a `docs/examples` Python script after receiving feedback.

### CODE REVIEW

File: src/neurotechdevkit/imaging/beamform.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    start_time: float = 0,  # s
+    speed_sound: float = 1540,  # m/s
+    f_number: Optional[float] = 1.5,
+    width: Optional[float] = None,  # m
+    bandwidth: Optional[float] = 0.75,
+    method: InterpolationMethod = InterpolationMethod.LINEAR,
+) -> csr_array:
+    """
+    Calculate the delay-and-sum sparse matrix for beamforming I/Q signals.
+
```
Comment (NewtonSander) [inline src/neurotechdevkit/imaging/beamform.py:119]: could you add the `Args: ` section?

File: src/neurotechdevkit/imaging/beamform.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        # Eq. 13 of https://doi.org/10.1016/j.ultras.2020.106309
+        result = minimize_scalar(
+            lambda theta: np.abs(
+                directivity(theta, width, wavelength_min) - optimal_directivity
+            ),
+            bounds=(0, np.pi / 2),
+        )
+        alpha = result.x
+        f_number = 1 / (2 * np.tan(alpha))
```
Comment (NewtonSander) [inline src/neurotechdevkit/imaging/beamform.py:172]: does it make sense to move this to another function?

File: src/neurotechdevkit/imaging/beamform.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+    # Beamform
+    iq_signals_column_vec = iq_signals.reshape(-1, 1)
+    beamformed_iq_signals = das_matrix @ iq_signals_column_vec
+    beamformed_iq_signals = beamformed_iq_signals.reshape(x.shape)
+
+    return beamformed_iq_signals
+
+
+def delay_and_sum_matrix(
```
Comment (NewtonSander) [inline src/neurotechdevkit/imaging/beamform.py:100]: I wonder if splitting this function into various smaller parts performing smaller steps makes sense. Would it become more readable? or the way it's currently written is better?

**Comment by d-lucena** [comment]
Thanks for the great new functionality, @charlesincharge!
I thought the notebooks are great as they are. I'd agree with Newton on reducing some of the complexity the main DAS function would be worthwhile.

For validation/integration testing, we could add a notebook that does the comparison as you suggested (and as we discussed doing for the stimulation results as well). This could be a separate story/PR though.

File: src/neurotechdevkit/imaging/beamform.py (hunk 1 above)
Comment (charlesincharge) [reply src/neurotechdevkit/imaging/beamform.py:138]: Added!

File: src/neurotechdevkit/imaging/beamform.py (hunk 2 above)
Comment (charlesincharge) [reply src/neurotechdevkit/imaging/beamform.py]: Yes, updated! See `_calculate_receptive_fields`

File: src/neurotechdevkit/imaging/beamform.py (hunk 3 above)
Comment (charlesincharge) [reply src/neurotechdevkit/imaging/beamform.py:100]: Updated!

**Comment by charlesincharge** [comment]
Seems like the `stride` installation is failing in `docker-test`. Perhaps this has to do with conflicting dependencies with `xarray`? Anyone have guidance on this? Perhaps we could move `stride` into the `poetry` dependencies?

**Comment by d-lucena** [comment]
Thanks @charlesincharge! The notebooks look good to me, thank you! I made some minor changes to keep it consistent across the two, but feel free to revert or change something I missed or is not correct.
Could you transfer them to a python file so they're rendered in the documentation gallery, please?

Regarding adding `stride` into `poetry` -- I believe that causes an issue with the `pip` deployment of the package; @NewtonSander, can you confirm?

**Comment by charlesincharge** [comment]
Thanks @d-lucena ! Also, looks like the tests are now passing; I wonder if the `stride` installation issue was related to https://github.com/agencyenterprise/neurotechdevkit/issues/149 and now fixed.

**Pull request merged**